src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.Exchange.cs

```diff
@@ -405,6 +417,28 @@ public ValueTask HandleAsync(ExtractTokenResponseContext context)
                     context.Response.ExpiresIn = value;
                 }
 
+                // Note: Huawei returns a non-standard "error" parameter as a numeric value, which is not allowed
+                // by OpenIddict (that requires a string). Huawei also returns a non-standard "sub_error" parameter
+                // that contains additional error information, with which the error code can demonstrate a specific
+                // meaning. To work around that, the "error" parameter is replaced with a standard error code.
+                // When the error code is "1101", the sub-error code of "20411" indicates that the device code
+                // authorization request is still waiting for the user to access the authorization page, and the
+                // sub-error code of "20412" indicates that the user has not performed the device code authorization.
+                // For more information about the error codes, sub-error codes, and their meanings, see:
+                // https://developer.huawei.com/consumer/en/doc/HMSCore-Guides/open-platform-error-0000001053869182#section6581130161218
+                else if (context.Registration.ProviderType is ProviderTypes.Huawei &&
+                         (JsonElement?) context.Response[Parameters.Error] is { ValueKind: JsonValueKind.Number } &&
+                         (JsonElement?) context.Response["sub_error"] is { ValueKind: JsonValueKind.Number })
+                {
+                    long error = (long) context.Response[Parameters.Error];
+                    long subError = (long) context.Response["sub_error"];
+                    context.Response[Parameters.Error] = (error, subError) switch
+                    {
+                        (1101, 20411 or 20412) => Errors.AuthorizationPending,
```

Comment:
Their documentation isn't clear at all about that, but is there an error code/sub error code that indicates that the device code or user code is expired (in this case, the user is no longer able to approve the authorization demand)?

If so, that would be nice to map it to the standard `expired_token` error.

Reply:
Here is the response payload when the user code expires:

![image](https://github.com/openiddict/openiddict-core/assets/21241496/f1f76405-71b9-4378-bf54-b32b4795da21)

So let me map 1101-20404 to `Errors.ExpiredToken`.

Reply:
![image](https://github.com/openiddict/openiddict-core/assets/21241496/8c0da257-9b35-4acd-b322-43d1e8fa3f7b)

Reply:
> So let me map 1101-20404 to `Errors.ExpiredToken`.

👍🏻 



> ![image](https://private-user-images.githubusercontent.com/21241496/340102962-8c0da257-9b35-4acd-b322-43d1e8fa3f7b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg1NDM4MzcsIm5iZiI6MTcxODU0MzUzNywicGF0aCI6Ii8yMTI0MTQ5Ni8zNDAxMDI5NjItOGMwZGEyNTctOWIzNS00YWNkLWIzMjItNDNkMWU4ZmEzZjdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE2VDEzMTIxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyMDFhMmJmYTExNWZkYTgxODYwMDlhODdmYTM1YTIxYWE1ZTZkYWIwMmI3OTk4NGQzZTU5M2ZkYTQ2NzVmNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.z5oQqntvxLgqOC6fPYOBQKQKWsOXH986UEdLJQ3JJYc)

Hum, they also have this one. Did you see it being returned?

![image](https://github.com/openiddict/openiddict-core/assets/6998306/308f270e-c18b-48f4-a6df-8622c967d66f)

Reply:
Done.